### ISSUE TITLE
File handle leak.

### ISSUE THREAD

**Issue by JontomXire** [gameplay3d/GamePlay#1858]
I have an admittedly slightly old fork, but it is leaking file handles like there is no tomorrow. I am not sure if there is a leak in Image::create(don't forget to fix error cases too) or not. Some similar code closes the stream, some just lets the "stream" object drop out of scope where, possibly (I'm not an expert here) the unique_ptr calls delete on the pointer it holds which also closes the file.

Further experimentation shows that the file handles appear to be leaked during the rendering process, while the game is running.

**Comment by JontomXire** [comment]
strace shows repeated sets of:

stat("/dev/input/js0", {st_mode=S_IFCHR|0664, st_rdev=makedev(13, 0), ...}) = 0
stat("/dev/input/js1", {st_mode=S_IFCHR|0664, st_rdev=makedev(13, 1), ...}) = 0
open("/sys/class/input/js1/device/id/vendor", O_RDONLY) = 306
read(306, "80ee\n", 8191)               = 5
close(306)                              = 0
open("/sys/class/input/js1/device/id/product", O_RDONLY) = 306
read(306, "cafe\n", 8191)               = 5
close(306)                              = 0
open("/dev/input/js1", O_RDONLY|O_NONBLOCK) = 306
fcntl(306, F_GETFL)                     = 0x8800 (flags O_RDONLY|O_NONBLOCK|O_LARGEFILE)
ioctl(306, JSIOCGNAME(256), 0x7ffcbeb295d0) = 29
ioctl(306, JSIOCGAXES, 0x7ffcbeb295ce)  = 0
ioctl(306, JSIOCGBUTTONS, 0x7ffcbeb295cf) = 0
stat("/dev/input/js2", 0x7ffcbeb297c0)  = -1 ENOENT (No such file or directory)
stat("/dev/input/js3", 0x7ffcbeb297c0)  = -1 ENOENT (No such file or directory)
stat("/dev/input/js4", 0x7ffcbeb297c0)  = -1 ENOENT (No such file or directory)
stat("/dev/input/js5", 0x7ffcbeb297c0)  = -1 ENOENT (No such file or directory)
stat("/dev/input/js6", 0x7ffcbeb297c0)  = -1 ENOENT (No such file or directory)
stat("/dev/input/js7", 0x7ffcbeb297c0)  = -1 ENOENT (No such file or directory)
stat("/dev/input/js8", 0x7ffcbeb297c0)  = -1 ENOENT (No such file or directory)
stat("/dev/input/js9", 0x7ffcbeb297c0)  = -1 ENOENT (No such file or directory)
stat("/dev/input/js10", 0x7ffcbeb297c0) = -1 ENOENT (No such file or directory)
stat("/dev/input/js11", 0x7ffcbeb297c0) = -1 ENOENT (No such file or directory)
stat("/dev/input/js12", 0x7ffcbeb297c0) = -1 ENOENT (No such file or directory)
stat("/dev/input/js13", 0x7ffcbeb297c0) = -1 ENOENT (No such file or directory)
stat("/dev/input/js14", 0x7ffcbeb297c0) = -1 ENOENT (No such file or directory)
stat("/dev/input/js15", 0x7ffcbeb297c0) = -1 ENOENT (No such file or directory)

I'm not sure why we keep stat'ing files that we already know don't exist.

After we open "/dev/input/js1" we never close it again. The file handle is leaked.

**Comment by JontomXire** [comment]
Multiple returns is a really bad coding practice. It would be a good idea for you to get a code quality checker and work on removing them.

In handleConnectedGamepad() in PlatformLinux.cpp there are a couple of returns where the file handle is not closed first. The function to enumerate devices is called once per frame, so if a device doesn't measure up the function returns, leaking the file handle, only to be called again and leak a new file handle a few milliseconds later. Apart from fixing the leak there are some performance optimisations, like only calling enumGamepads() once every half second or so.

**Comment by seanpaultaylor** [comment]
Please submit a pull request if you have any fix for the file handle leak or any optimizations.

Generally I agree good to "try" to have as few returns as possible. 
But not at the cost of poor code readability. Here is a good thread on the topic.
http://stackoverflow.com/questions/36707/should-a-function-have-only-one-return-statement

I am always open to any changes that can improve the code quality and performance optimizations.

**Comment by JontomXire** [comment]
I don't have an exact fork of your repo. Rather I imported it into a
new repo that includes the dependencies as well, and then made a lot
of customisation. I think I gave you enough information to be able to
fix it and it is a really quick fix.

Incidentally, without the fix it was taking around 10 seconds for my
game to exit. With the fix it is instantaneous.
 Sent using Hushmail
On 1/27/2017 at 3:51 PM, "Sean Taylor"  wrote:

	Please submit a pull request if you have any fix for the file handle
leak or any optimizations. 

	Generally I agree good to "try" to have as few returns as possible.
 But not at the cost of poor code readability. Here is a good thread
on the topic.
 http://stackoverflow.com/questions/36707/should-a-function-have-only-one-return-statement


	I am always open to any changes that can improve the code quality and
performance optimizations.
